In [2]:
import numpy as np
import pandas as pd
import  datetime as dt


In [3]:
!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ./data/

--2022-12-01 16:36:35--  https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt
Resolving www.dropbox.com (www.dropbox.com)... 162.125.80.18, 2620:100:6035:18::a27d:5512
Connecting to www.dropbox.com (www.dropbox.com)|162.125.80.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: /s/raw/xvjzaxzz3ysphme/data_000637.txt [following]
--2022-12-01 16:36:35--  https://www.dropbox.com/s/raw/xvjzaxzz3ysphme/data_000637.txt
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc387eaf0f22f30c91e1f55f6e0d.dl.dropboxusercontent.com/cd/0/inline/BxyxvUwwLzYoGjBC-ub-FXp3_kH3B0YUxjffHLpcjE0Q--vUSldl2mnlJL9WoqM12wpEQ5kyUlCKIVXIqYPB-atX7cvM8f9U85d3zYvkgiilKlp8IOK6ktvouN3GZuz_du7oC4nMwI55J2wF-wFd50WitjLEHu-eLjkX4iSyWMLMdw/file# [following]
--2022-12-01 16:36:36--  https://uc387eaf0f22f30c91e1f55f6e0d.dl.dropboxusercontent.com/cd/0/inline/BxyxvUwwLzYoGjBC-ub-FXp3_kH3B0YUxjffHLpcjE0Q--vUSldl2mnlJL9WoqM12wpE

In [4]:
file_name = "./data/data_000637.txt"
data = pd.read_csv(file_name)
data


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0           1     0          123  3869200167        2374        26
1           1     0          124  3869200167        2374        27
2           1     0           63  3869200167        2553        28
3           1     0           64  3869200167        2558        19
4           1     0           64  3869200167        2760        25
...       ...   ...          ...         ...         ...       ...
1310715     1     0           62  3869211171         762        14
1310716     1     1            4  3869211171         763        11
1310717     1     0           64  3869211171         764         0
1310718     1     0          139  3869211171         769         0
1310719     1     0           61  3869211171         762        18

[1310720 rows x 6 columns]

1. Create a Pandas DataFrame reading N rows of the data/data_000637.txt dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k (check the documentation).

ex 1
the maximum number of rows 

In [5]:
print(len(data))
print(data.shape)
print(data.axes[0])

1310720
(1310720, 6)
RangeIndex(start=0, stop=1310720, step=1)


 we choose n to be smaller than 1310720 and larger than 10000, so we choose 999999

In [6]:
df=data.sample(n=999999)
df.shape

(999999, 6)

2. Estimate the number of BX in a ORBIT (the value x).
Hint: check when the BX counter reaches the maximum value before being reset to 0.

In [7]:
max=df["BX_COUNTER"].max()
min=df['BX_COUNTER'].min()
print(max)
print(min)

3563
0


so, the maximum number in 'BX_COUNTER' column is 3563 so after that it reaches at minimum which is zero
so each orbit is 3653 of BX.

3. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information) since the beginning of the data acquisition, and convert the new column to a Time Series.

In [8]:
print(25/30)
t=25/30
s=25*3563
df['newcolumn']=(df['TDC_MEAS']*t)+(df['BX_COUNTER']*25)+(df['ORBIT_CNT']*s)
df


0.8333333333333334


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
875299      1     0           64  3869208178        3521        13   
1185634     1     0          139  3869210326        3406         0   
213385      1     1            3  3869202291        1710        12   
195734      1     1            4  3869202114        3000        12   
225940      1     0          123  3869202415        2776         4   
...       ...   ...          ...         ...         ...       ...   
146815      1     0           56  3869201628        2426         9   
441810      1     0            2  3869204621        1900        24   
14806       1     0           64  3869200315        1339        10   
951655      1     0           64  3869208705         641        13   
632544      1     1           64  3869206531        2699        11   

            newcolumn  
875299   3.446497e+14  
1185634  3.446499e+14  
213385   3.446492e+14  
195734   3.446492e+14  
225940   3.446492e+14  
...               ...  
146815   3.446491e+14  
441810   3.446494e+14  
14806    3.446490e+14  
951655   3.446498e+14  
632544   3.446496e+14  

[999999 rows x 7 columns]

In [9]:
# To define a date, the datetime module is very useful
import datetime as dt

date = dt.date.today()
date

datetime.date(2022, 12, 1)

In [10]:
# Get the timestamp, which is the nanoseconds from January 1st 1970
tstamp = pd.Timestamp(date)
#tstamp = pd.Timestamp(dt.datetime(1970, 1, 1, 0, 0, 0, 1))
print("Timestamp:", tstamp.value)
print(pd.Timestamp(tstamp.value,unit='ns'))

Timestamp: 1669852800000000000
2022-12-01 00:00:00


In [35]:
def convertns_totimestamp(x):
  return pd.Timestamp(x,unit='ns')
df['newcolumn']=df['newcolumn'].apply(convertns_totimestamp)
print(df.iloc[1,:])
print(df.head())

HEAD                                       1
FPGA                                       0
TDC_CHANNEL                              139
ORBIT_CNT                         3869210326
BX_COUNTER                              3406
TDC_MEAS                                   0
newcolumn      1970-01-04 23:44:09.909873600
Name: 1185634, dtype: object
         HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
875299      1     0           64  3869208178        3521        13   
1185634     1     0          139  3869210326        3406         0   
213385      1     1            3  3869202291        1710        12   
195734      1     1            4  3869202114        3000        12   
225940      1     0          123  3869202415        2776         4   

                            newcolumn  
875299  1970-01-04 23:44:09.718543385  
1185634 1970-01-04 23:44:09.909873600  
213385  1970-01-04 23:44:09.194113585  
195734  1970-01-04 23:44:09.178379560  
225940  1970-01-04 23:44:09.2051855

4. Find out the duration of the data taking in hours, minutes and seconds, by using the features of the Time Series. Perform this check reading the whole dataset.



In [36]:
df2=df.copy()
print(df2.sort_values(by='newcolumn')['newcolumn'])
orderColumn=df2.sort_values(by='newcolumn')['newcolumn']
#first=df2['newcolumn'][0]
#last=df2['newcolumn'][999999]
#print('first:',first)
#print('last',last)
#diff=last-first
print("df2")
print(orderColumn)
print('orderColumn shape :',orderColumn.shape)
min=list(orderColumn)[0]
max=list(orderColumn)[999998]
print('min :',min)
print('max :',max)
print("the difference is :", max-min)

2         1970-01-04 23:44:09.004939373
3         1970-01-04 23:44:09.004939490
4         1970-01-04 23:44:09.004944545
5         1970-01-04 23:44:09.004944578
8         1970-01-04 23:44:09.004944892
                       ...             
1310714   1970-01-04 23:44:09.985075876
1310715   1970-01-04 23:44:09.985075886
1310719   1970-01-04 23:44:09.985075890
1310716   1970-01-04 23:44:09.985075909
1310718   1970-01-04 23:44:09.985076050
Name: newcolumn, Length: 999999, dtype: datetime64[ns]
df2
2         1970-01-04 23:44:09.004939373
3         1970-01-04 23:44:09.004939490
4         1970-01-04 23:44:09.004944545
5         1970-01-04 23:44:09.004944578
8         1970-01-04 23:44:09.004944892
                       ...             
1310714   1970-01-04 23:44:09.985075876
1310715   1970-01-04 23:44:09.985075886
1310719   1970-01-04 23:44:09.985075890
1310716   1970-01-04 23:44:09.985075909
1310718   1970-01-04 23:44:09.985076050
Name: newcolumn, Length: 999999, dtype: datetime64[ns]
orderC

5. Use the .groupby() method to find out the noisy channels, i.e. the TDC channels with most counts (print to screen the top 3 and the corresponding counts)



first check how different values are in 'TDC-CHANNEL' column are.

In [37]:
numberOfUniqueValues=df['TDC_CHANNEL'].unique()
print(numberOfUniqueValues)
print('the number of unique values are',numberOfUniqueValues.shape[0])

[ 64 139   3   4 123  13  16  31   2  58  56  62  63  61  40  29  54  37
 121  51   8  60   6  57 124  14 114 104  44  12 120  33  41   7   1  19
  84  47  45  66  59  42  73  50  76   5  36  69  46  86 116 112 115 113
  28  72  10 122 118  24  15  85 107 119  90  48  38  78  53  20  22  49
  68  75  79  32  11  91  30  82  27  26 117 108 101   9  83  17  99  92
 110  70  23  74 106  80  18  21  93  98  77 128 125 111  52  87  81  35
  65  43  71  96  39 103  34 100  89  25 127  55  94  95 105 102  97  88
  67 137 109 126 130 138 129]
the number of unique values are 133


In [38]:
noisy_channels=df.groupby('TDC_CHANNEL').count()
print(noisy_channels)

              HEAD   FPGA  ORBIT_CNT  BX_COUNTER  TDC_MEAS  newcolumn
TDC_CHANNEL                                                          
1            22777  22777      22777       22777     22777      22777
2            26225  26225      26225       26225     26225      26225
3            18011  18011      18011       18011     18011      18011
4            21961  21961      21961       21961     21961      21961
5            12503  12503      12503       12503     12503      12503
...            ...    ...        ...         ...       ...        ...
129             31     31         31          31        31         31
130             59     59         59          59        59         59
137             47     47         47          47        47         47
138             58     58         58          58        58         58
139          82339  82339      82339       82339     82339      82339

[133 rows x 6 columns]


In [39]:
#now we sort the previous table to get the most noisy channel
order_noisy=noisy_channels.sort_values(by='HEAD',ascending=False)
print(order_noisy)
#now we have the count of each TCD_CHANNEL count and we sort them so, the first three rows of the order_noisy table is the solution
print("rint to screen the top 3 and the corresponding counts \n",order_noisy.iloc[:3,:])
print("the corresponding values for order in TDC_CHANNEL are \n\n",order_noisy.iloc[:3,:].index )

              HEAD   FPGA  ORBIT_CNT  BX_COUNTER  TDC_MEAS  newcolumn
TDC_CHANNEL                                                          
139          82339  82339      82339       82339     82339      82339
64           50478  50478      50478       50478     50478      50478
63           49262  49262      49262       49262     49262      49262
62           37841  37841      37841       37841     37841      37841
61           37794  37794      37794       37794     37794      37794
...            ...    ...        ...         ...       ...        ...
126            465    465        465         465       465        465
130             59     59         59          59        59         59
138             58     58         58          58        58         58
137             47     47         47          47        47         47
129             31     31         31          31        31         31

[133 rows x 6 columns]
rint to screen the top 3 and the corresponding counts 
           

6.Count the number of non-empty orbits (i.e. the number of orbits with at least one hit)

In [51]:
c=df['ORBIT_CNT'].unique().size
print(c)
print(df['ORBIT_CNT'].isnull().sum())
print('there is 0 null value so :',c-(df['ORBIT_CNT'].isnull().sum()))

11001
0
there is 0 null value so : 11001


7. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139.



In [52]:
query1 = df.query('TDC_CHANNEL==139')
print(query1)
print(query1.shape)
print('unique orbit')
uniqeOrbitCount=query1['ORBIT_CNT'].unique().shape
print(uniqeOrbitCount)

         HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
1185634     1     0          139  3869210326        3406         0   
742852      1     0          139  3869207291        2979         0   
1216283     1     0          139  3869210534        2555         0   
968983      1     0          139  3869208811         290         0   
189250      1     0          139  3869202050        2760         0   
...       ...   ...          ...         ...         ...       ...   
414217      1     0          139  3869204337        2706         0   
337238      1     0          139  3869203544        3040         0   
1243093     1     1          139  3869210720        1237         0   
681655      1     0          139  3869206856         424         0   
1261306     1     0          139  3869210834        1676         0   

                            newcolumn  
1185634 1970-01-04 23:44:09.909873600  
742852  1970-01-04 23:44:09.639520300  
1216283 1970-01-04 23:44:09.928379925  

8. Create two Series (one for each FPGA) that have the TDC channel as index, and the number of counts for the corresponding TDC channel as values.



In [62]:
h=df.query('FPGA==0')
print('the data with FPGA==0')
print(h)
print("now with value_counts we obtain the count of each unique TDC_CHANNEL ")
print('with this method it has the TDC_CHANNEL channel as index itself')
l=h['TDC_CHANNEL'].value_counts()
print(l)
print('the index are the TDC_CHANNEL unique values according to the question')
#print(l.index)
#print(l.index.shape)

the data with FPGA==0
         HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
875299      1     0           64  3869208178        3521        13   
1185634     1     0          139  3869210326        3406         0   
225940      1     0          123  3869202415        2776         4   
1056698     1     0           13  3869209423        2700        18   
742852      1     0          139  3869207291        2979         0   
...       ...   ...          ...         ...         ...       ...   
222325      1     0           51  3869202379        2609        29   
146815      1     0           56  3869201628        2426         9   
441810      1     0            2  3869204621        1900        24   
14806       1     0           64  3869200315        1339        10   
951655      1     0           64  3869208705         641        13   

                            newcolumn  
875299  1970-01-04 23:44:09.718543385  
1185634 1970-01-04 23:44:09.909873600  
225940  1970-01-0

In [61]:
h2=df.query('FPGA==1')
print("now with value_counts we obtain the count of each unique TDC_CHANNEL ")
print('with this method it has the TDC_CHANNEL channel as index itself')
l2=h2['TDC_CHANNEL'].value_counts()
print(type(l2))
print(l2)
print('the index are the TDC_CHANNEL unique values according to the question')
#print(l2.index)
#print(l2.index.shape)

now with value_counts we obtain the count of each unique TDC_CHANNEL 
with this method it has the TDC_CHANNEL channel as index itself
<class 'pandas.core.series.Series'>
2      24990
139    24836
1      21855
4      20165
3      16858
       ...  
9         61
130       33
138       31
129       29
137       24
Name: TDC_CHANNEL, Length: 132, dtype: int64
the index are the TDC_CHANNEL unique values according to the question
